packages/web/nexus/subscription.ts

```diff
@@ -0,0 +1,322 @@
+import { MembershipSubscriptionPeriod, InputJsonValue, PrismaClient } from '@journaly/j-db-client'
+import {
+  arg,
+  booleanArg,
+  extendType,
+  objectType,
+  stringArg,
+} from '@nexus/schema'
+import stripe from './utils/stripe'
+
+const MembershipSubscription = objectType({
+  name: 'MembershipSubscription',
+  definition(t) {
+    t.model.id()
+    t.model.period()
+    t.model.userId()
+    t.model.expiresAt()
+    t.model.cancelAtPeriodEnd()
+    t.model.lastFourCardNumbers()
+    t.model.cardType()
+  }
+})
+
+const getSubscriptionPriceId = (subType: MembershipSubscriptionPeriod) => {
+  switch(subType) {
+    case MembershipSubscriptionPeriod.MONTHLY:
+      return 'price_1ISRgvB8OEjVdGPaQr7ZANW8'
+    case MembershipSubscriptionPeriod.QUARTERLY:
+      return 'price_1ISRgvB8OEjVdGPaeOx4m255'
+    case MembershipSubscriptionPeriod.ANNUALY:
+      return 'price_1ISRgvB8OEjVdGPam1PTr6hE'
+  }
+}
+
+const setPaymentMethod = async (
+  userId: number,
+  db: PrismaClient,
+  customerId: string,
+  paymentMethodId: string,
+  lastFourCardNumbers: string = '1234',
+  cardType: string = 'Visa'
+) => {
+  const customer = await stripe.customers.retrieve(customerId)
+  
+  if (!customer) throw new Error("User has stripeCustomerId but unable to find customer in Stripe")
+
+  await stripe.paymentMethods.attach(paymentMethodId, {
+    customer: customer.id,
+  })
+
+  // Update customer's default method in case they've entered a new card
+  await stripe.customers.update(customer.id, {
+    invoice_settings: {
+      default_payment_method: paymentMethodId,
+    },
+  })
+
+  return db.membershipSubscription.update({
+    where: {
+      userId,
+    },
+    data: {
+      lastFourCardNumbers,
+      cardType,
+    },
+  })
+}
+
+const setPlan = async (
+  userId: number,
+  db: PrismaClient,
+  stripeSubscriptionId: string,
+  subscriptionPeriod: MembershipSubscriptionPeriod,
+  cancelAtPeriodEnd: boolean,
+) => {
+  const stripeSubscription = await stripe.subscriptions.retrieve(stripeSubscriptionId)
+  const subscriptionUpdated = await stripe.subscriptions.update(stripeSubscription.id, {
+    payment_behavior: 'pending_if_incomplete',
+    proration_behavior: 'always_invoice',
+    items: [{
+      id: stripeSubscription.items.data[0].id,
+      price: getSubscriptionPriceId(subscriptionPeriod),
+    }],
+  })
+  if (subscriptionUpdated.pending_update) {
+    // Payment failed
+    throw new Error("Unable to update subscription, possible payment failure")
+  }
+  await stripe.subscriptions.update(stripeSubscription.id, {
+    cancel_at_period_end: cancelAtPeriodEnd,
+  })
+  // Update our records with the new subscription info
+  return db.membershipSubscription.update({
+    where: {
+      userId,
+    },
+    data: {
+      period: subscriptionPeriod,
+      expiresAt: new Date(subscriptionUpdated.current_period_end * 1000 + (24 * 60 * 60 * 1000 * 2)),
+      stripeSubscription: subscriptionUpdated as unknown as InputJsonValue,
```

Comment:
I think we also want to update `cancelAtPeriodEnd` in this DB call

Reply:
Great catch!